example/src/main/java/io/netty/example/http2/tiles/Http2RequestHandler.java

```diff
@@ -0,0 +1,115 @@
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.QueryStringDecoder;
+import io.netty.handler.codec.http2.HttpUtil;
+import io.netty.handler.codec.http2.InboundHttp2ToHttpAdapter;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Handles all the requests for data. It receives a {@link FullHttpRequest},
+ * which has been converted by a {@link InboundHttp2ToHttpAdapter} before it
+ * arrived here. For further details, check {@link Http2OrHttpHandler} where the
+ * pipeline is setup.
+ */
+public class Http2RequestHandler extends SimpleChannelInboundHandler<FullHttpRequest> {
+
+    private static final String LATENCY = "latency";
+    private static final String IMAGE_COORDINATE_Y = "y";
+    private static final String IMAGE_COORDINATE_X = "x";
+
+    @Override
+    protected void channelRead0(ChannelHandlerContext ctx, FullHttpRequest request) throws Exception {
+        QueryStringDecoder queryString = new QueryStringDecoder(request.uri());
+        String streamId = streamId(request);
+        int latency = latency(queryString);
+        if (missing(queryString, IMAGE_COORDINATE_X) || missing(queryString, IMAGE_COORDINATE_Y)) {
+            handlePage(ctx, streamId, latency);
+        } else {
+            handleImage(queryString, ctx, streamId, latency);
+        }
+    }
+
+    private void handleImage(QueryStringDecoder query, ChannelHandlerContext ctx, String streamId, int latency) {
+        int x = toInt(firstValue(query, IMAGE_COORDINATE_X), 0);
+        int y = toInt(firstValue(query, IMAGE_COORDINATE_Y), 0);
+        byte[] image = ImageCache.INSTANCE.image(x, y);
+        FullHttpResponse response = new DefaultFullHttpResponse(HTTP_1_1, OK, wrappedBuffer(image));
+        response.headers().set(CONTENT_TYPE, "image/jpeg");
+        sendResponse(ctx, streamId, latency, response);
+    }
+
+    private void handlePage(ChannelHandlerContext ctx, String streamId, int latency) {
+        ByteBuf content = Unpooled.wrappedBuffer(Html.HEADER, Html.body(latency), Html.FOOTER);
+        FullHttpResponse response = new DefaultFullHttpResponse(HTTP_1_1, OK, content);
+        response.headers().set(CONTENT_TYPE, "text/html; charset=UTF-8");
+        sendResponse(ctx, streamId, latency, response);
+    }
+
+    protected void sendResponse(final ChannelHandlerContext ctx, String streamId, int latency,
+            final FullHttpResponse response) {
+        setContentLength(response, response.content().readableBytes());
+        streamId(response, streamId);
+        ctx.executor().schedule(new Runnable() {
+            @Override
+            public void run() {
+                ctx.writeAndFlush(response);
+            }
+        }, latency, TimeUnit.MILLISECONDS);
+    }
+
+    private int latency(QueryStringDecoder queryString) {
+        int latency = toInt(firstValue(queryString, LATENCY), 0);
+        // Make sure that latency is not too big
+        return abs(latency > 1000 ? 0 : latency);
```

Comment:
Instead of allowing invalid input and translating it to valid input on the backend...can we just validate the request and return a `400` if it is not valid?